File: mage/src/main/java/mil/nga/giat/mage/sdk/gson/deserializer/RolesDeserializer.java
package mil.nga.giat.mage.sdk.gson.deserializer;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Collection;

import mil.nga.giat.mage.sdk.datastore.user.Role;

/**
 * JSON to {@link Role}
 * 
 * @author newmanw
 * 
 */
public class RolesDeserializer implements JsonDeserializer<Collection<Role>> {

	private static final String LOG_NAME = RolesDeserializer.class.getName();

	private Gson roleDeserializer;

	public RolesDeserializer() {
		roleDeserializer = RoleDeserializer.getGsonBuilder();
	}

	/**
	 * Convenience method for returning a Gson object with a registered GSon
	 * TypeAdaptor i.e. custom deserializer.
	 * 
	 * @return A Gson object that can be used to convert Json into a {@link Role}.
	 */
	public static Gson getGsonBuilder() {
		GsonBuilder gsonBuilder = new GsonBuilder();
		gsonBuilder.registerTypeAdapter(new TypeToken<Collection<Role>>(){}.getType(), new RolesDeserializer());
		return gsonBuilder.create();
	}

	@Override
	public Collection<Role> deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
		Collection<Role> roles = new ArrayList<>();

		JsonArray jsonRoles = json.getAsJsonArray();
		for (JsonElement element : jsonRoles) {
			Role role = roleDeserializer.fromJson(element, Role.class);
			roles.add(role);
		}

		return roles;
	}
}

File: mage/src/main/java/mil/nga/giat/mage/network/role/RolesDeserializer.kt
package mil.nga.giat.mage.network.role

import android.util.Log
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import mil.nga.giat.mage.sdk.datastore.user.Permission
import mil.nga.giat.mage.database.model.permission.Role
import mil.nga.giat.mage.sdk.datastore.user.Permissions
import java.lang.reflect.Type

class RolesDeserializer : JsonDeserializer<List<Role>> {

   @Throws(JsonParseException::class)
   override fun deserialize(
      json: JsonElement,
      typeOfT: Type,
      context: JsonDeserializationContext
   ): List<Role> {
      val roles = mutableListOf<Role>()
      val jsonRoles = json.asJsonArray
      for (element in jsonRoles) {
         val role = deserializeRole(element)
         roles.add(role)
      }
      return roles
   }

   private fun deserializeRole(json: JsonElement): Role {
      val jsonRole = json.asJsonObject
      val remoteId = jsonRole["id"].asString
      val name = jsonRole["name"].asString
      val description = jsonRole["description"].asString
      val permissions = mutableListOf<Permission>()

      jsonRole["permissions"].asJsonArray.forEach { element ->
         element?.asString?.let { jsonPermission ->
            try {
               val permission = Permission.valueOf(jsonPermission.uppercase())
               permissions.add(permission)
            } catch (iae: IllegalArgumentException) {
               Log.e(LOG_NAME, "Could not find matching permission, $jsonPermission, for user.")
            }
         }
      }

      return Role(remoteId, name, description, Permissions(permissions))
   }

   companion object {
      private val LOG_NAME = RolesDeserializer::class.java.name
   }
}